File: examples/js/stream_client.mjs
#!/usr/bin/env node
/**
 * Streaming transcription client for HighPerfASR (Node.js).
 *
 * Opens a WebSocket to the streaming endpoint, sends PCM16 audio chunks
 * from a WAV file, prints partial transcripts as they arrive, and prints
 * the final transcript on close.
 *
 * Dependencies: npm install ws
 *
 * Usage:
 *   node stream_client.mjs audio.wav
 *   node stream_client.mjs --url ws://myserver:8001 audio.wav
 *   node stream_client.mjs --chunk-ms 200 audio.wav
 */

import { readFileSync } from "fs";
import { basename } from "path";
import WebSocket from "ws";

const SAMPLE_RATE = 16000;
const SAMPLE_WIDTH = 2; // 16-bit PCM = 2 bytes per sample
const DEFAULT_CHUNK_MS = 100;

/**
 * Parse a WAV file and return the raw PCM data (skips the header).
 * Validates basic format properties and warns on mismatches.
 */
function readPcm16FromWav(filePath) {
  const buf = readFileSync(filePath);

  // Minimal WAV header parsing
  const riff = buf.toString("ascii", 0, 4);
  const wave = buf.toString("ascii", 8, 12);
  if (riff !== "RIFF" || wave !== "WAVE") {
    throw new Error(`${filePath} is not a valid WAV file`);
  }

  // Walk chunks to find "fmt " and "data"
  let offset = 12;
  let fmtFound = false;
  let channels, sampleRate, bitsPerSample, dataStart, dataSize;

  while (offset < buf.length - 8) {
    const chunkId = buf.toString("ascii", offset, offset + 4);
    const chunkSize = buf.readUInt32LE(offset + 4);

    if (chunkId === "fmt ") {
      const audioFormat = buf.readUInt16LE(offset + 8);
      channels = buf.readUInt16LE(offset + 10);
      sampleRate = buf.readUInt32LE(offset + 12);
      bitsPerSample = buf.readUInt16LE(offset + 22);
      fmtFound = true;

      if (audioFormat !== 1) {
        console.warn(`Warning: WAV audio format is ${audioFormat}, expected 1 (PCM)`);
      }
    } else if (chunkId === "data") {
      dataStart = offset + 8;
      dataSize = chunkSize;
      break;
    }

    offset += 8 + chunkSize;
  }

  if (!fmtFound) throw new Error("WAV file missing fmt chunk");
  if (dataStart === undefined) throw new Error("WAV file missing data chunk");

  if (channels !== 1) {
    console.warn(`Warning: WAV has ${channels} channels, expected mono`);
  }
  if (sampleRate !== SAMPLE_RATE) {
    console.warn(`Warning: WAV sample rate is ${sampleRate} Hz, expected ${SAMPLE_RATE} Hz`);
  }
  if (bitsPerSample !== 16) {
    console.warn(`Warning: WAV is ${bitsPerSample}-bit, expected 16-bit`);
  }

  return buf.subarray(dataStart, dataStart + dataSize);
}

/**
 * Sleep for the given number of milliseconds.
 */
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Stream a WAV file over WebSocket and print transcripts.
 */
async function streamAudio(url, filePath, chunkMs) {
  const pcmData = readPcm16FromWav(filePath);
  const chunkBytes = Math.floor(SAMPLE_RATE * SAMPLE_WIDTH * (chunkMs / 1000));
  const endpoint = `${url.replace(/\/$/, "")}/v1/stream`;

  console.log(`Connecting to ${endpoint} ...`);

  return new Promise((resolve, reject) => {
    const ws = new WebSocket(endpoint);

    let opened = false;

    ws.on("message", (raw) => {
      const msg = JSON.parse(raw.toString());

      // Handle initial opened message
      if (!opened && msg.status === "opened") {
        opened = true;
        console.log(`Stream opened (id: ${msg.stream_id})`);

        // Start sending audio in the background
        (async () => {
          let offset = 0;
          while (offset < pcmData.length) {
            const chunk = pcmData.subarray(offset, offset + chunkBytes);
            ws.send(chunk);
            offset += chunk.length;
            await sleep(chunkMs);
          }
          // Signal end of audio
          ws.send(JSON.stringify({ action: "close" }));
        })();
        return;
      }

      // Error message
      if (msg.error) {
        console.error(`\nError: ${msg.error}`);
        ws.close();
        return;
      }

      // Final close message
      if (msg.status === "closed") {
        console.log(`\n\nFinal transcript: ${msg.final_text}`);
        ws.close();
        return;
      }

      // Partial transcript update
      const partial = msg.partial_transcript || "";
      const finalText = msg.final_transcript || "";
      if (finalText) {
        process.stdout.write(`\r[final]   ${finalText}\n`);
      }
      if (partial) {
        process.stdout.write(`\r[partial] ${partial}`);
      }
    });

    ws.on("close", () => {
      console.log("\nDone.");
      resolve();
    });

    ws.on("error", (err) => {
      console.error(`WebSocket error: ${err.message}`);
      reject(err);
    });
  });
}

// --- CLI argument parsing ---

function printUsage() {
  console.log(`Usage: node ${basename(process.argv[1])} [options] <file>

Options:
  --url <url>        WebSocket server URL (default: ws://localhost:8001)
  --chunk-ms <ms>    Chunk duration in ms (default: ${DEFAULT_CHUNK_MS})
  --help             Show this help message`);
}

function parseArgs() {
  const args = process.argv.slice(2);
  let url = "ws://localhost:8001";
  let chunkMs = DEFAULT_CHUNK_MS;
  let file = null;

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--url":
        url = args[++i];
        break;
      case "--chunk-ms":
        chunkMs = parseInt(args[++i], 10);
        break;
      case "--help":
      case "-h":
        printUsage();
        process.exit(0);
        break;
      default:
        if (args[i].startsWith("-")) {
          console.error(`Unknown option: ${args[i]}`);
          printUsage();
          process.exit(1);
        }
        file = args[i];
    }
  }

  if (!file) {
    console.error("Error: no audio file specified");
    printUsage();
    process.exit(1);
  }

  return { url, chunkMs, file };
}

const { url, chunkMs, file } = parseArgs();
streamAudio(url, file, chunkMs).catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/stream_client.py
#!/usr/bin/env python3
"""Streaming transcription client for HighPerfASR.

Opens a WebSocket to the streaming endpoint, sends PCM16 audio chunks
from a WAV file, prints partial transcripts as they arrive, and prints
the final transcript on close.

Usage:
    python stream_client.py audio.wav
    python stream_client.py --url ws://myserver:8001 audio.wav
    python stream_client.py --chunk-ms 200 audio.wav
"""

import argparse
import asyncio
import json
import sys
import wave

import websockets


SAMPLE_RATE = 16000
SAMPLE_WIDTH = 2  # 16-bit PCM = 2 bytes per sample
DEFAULT_CHUNK_MS = 100  # 100ms chunks


def read_pcm16_from_wav(file_path: str) -> bytes:
    """Read a WAV file and return raw PCM16 mono bytes.

    Validates format and warns on mismatches.
    """
    with wave.open(file_path, "rb") as wf:
        channels = wf.getnchannels()
        sample_width = wf.getsampwidth()
        framerate = wf.getframerate()
        n_frames = wf.getnframes()

        if sample_width != SAMPLE_WIDTH:
            print(
                f"Warning: WAV is {sample_width * 8}-bit, expected 16-bit. "
                f"Audio will be sent as-is.",
                file=sys.stderr,
            )
        if channels != 1:
            print(
                f"Warning: WAV has {channels} channels, expected mono. "
                f"Audio will be sent as-is.",
                file=sys.stderr,
            )
        if framerate != SAMPLE_RATE:
            print(
                f"Warning: WAV sample rate is {framerate} Hz, expected {SAMPLE_RATE} Hz. "
                f"Audio will be sent as-is.",
                file=sys.stderr,
            )

        pcm_data = wf.readframes(n_frames)

    return pcm_data


async def stream_audio(url: str, file_path: str, chunk_ms: int):
    """Stream a WAV file over WebSocket and print transcripts."""
    pcm_data = read_pcm16_from_wav(file_path)
    chunk_bytes = int(SAMPLE_RATE * SAMPLE_WIDTH * (chunk_ms / 1000))

    endpoint = f"{url.rstrip('/')}/v1/stream"
    print(f"Connecting to {endpoint} ...")

    async with websockets.connect(endpoint) as ws:
        # Step 1: Wait for the "opened" message
        msg = json.loads(await ws.recv())
        if msg.get("status") != "opened":
            print(f"Unexpected initial message: {msg}", file=sys.stderr)
            return
        stream_id = msg["stream_id"]
        print(f"Stream opened (id: {stream_id})")

        # Step 2: Send audio chunks while receiving partial transcripts
        async def send_audio():
            offset = 0
            total = len(pcm_data)
            while offset < total:
                chunk = pcm_data[offset : offset + chunk_bytes]
                await ws.send(chunk)
                offset += len(chunk)
                # Pace the send to approximate real-time playback
                await asyncio.sleep(chunk_ms / 1000)
            # Step 3: Signal end of audio
            await ws.send(json.dumps({"action": "close"}))

        async def recv_transcripts():
            while True:
                try:
                    raw = await ws.recv()
                except websockets.ConnectionClosed:
                    break
                msg = json.loads(raw)

                if "error" in msg:
                    print(f"\nError: {msg['error']}", file=sys.stderr)
                    break

                # Final close message
                if msg.get("status") == "closed":
                    print(f"\n\nFinal transcript: {msg['final_text']}")
                    break

                # Partial transcript update
                partial = msg.get("partial_transcript", "")
                final = msg.get("final_transcript", "")
                if final:
                    sys.stdout.write(f"\r[final]   {final}\n")
                if partial:
                    sys.stdout.write(f"\r[partial] {partial}")
                    sys.stdout.flush()

        # Run sender and receiver concurrently
        await asyncio.gather(send_audio(), recv_transcripts())

    print("\nDone.")


def main():
    parser = argparse.ArgumentParser(description="Streaming transcription client")
    parser.add_argument("file", help="Path to WAV file (16 kHz, mono, PCM16)")
    parser.add_argument(
        "--url",
        default="ws://localhost:8001",
        help="WebSocket server URL (default: ws://localhost:8001)",
    )
    parser.add_argument(
        "--chunk-ms",
        type=int,
        default=DEFAULT_CHUNK_MS,
        help=f"Chunk duration in milliseconds (default: {DEFAULT_CHUNK_MS})",
    )
    args = parser.parse_args()

    asyncio.run(stream_audio(args.url, args.file, args.chunk_ms))


if __name__ == "__main__":
    main()
